crates/biome_js_analyze/src/analyzers/nursery/no_duplicate_test_hooks.rs

```diff
@@ -0,0 +1,237 @@
+use biome_analyze::{
+    context::RuleContext, declare_rule, AddVisitor, Phases, QueryMatch, Queryable, Rule,
+    RuleDiagnostic, RuleSource, RuleSourceKind, ServiceBag, Visitor, VisitorContext,
+};
+use biome_console::markup;
+use biome_js_syntax::{AnyJsExpression, JsCallExpression, JsLanguage, JsSyntaxToken, TextRange};
+use biome_rowan::{AstNode, Language, SyntaxNode, WalkEvent};
+
+declare_rule! {
+    /// A `describe` block should not contain duplicate hooks.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// describe('foo', () => {
+    ///   beforeEach(() => {
+    ///     // some setup
+    ///   });
+    ///   beforeEach(() => {
+    ///     // some setup
+    ///   });
+    ///   test('foo_test', () => {
+    ///    // some test
+    ///   });
+    /// });
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// describe('foo', () => {
+    ///   beforeEach(() => {
+    ///     // some setup
+    ///   });
+    ///   test('foo_test', () => {
+    ///     afterAll(() => {
+    ///       // some teardown
+    ///     });
+    ///    afterAll(() => {
+    ///      // some teardown
+    ///    });
+    ///   });
+    /// });
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// describe('foo', () => {
+    ///   beforeEach(() => {
+    ///     // some setup
+    ///   });
+    ///   test('foo_test', () => {
+    ///     // some test
+    ///   });
+    /// });
+    /// ```
+    ///
+    pub(crate) NoDuplicateTestHooks {
+        version: "next",
+        name: "noDuplicateTestHooks",
+        recommended: true,
+        source: RuleSource::EslintJest("no-focused-tests"),
+        source_kind: RuleSourceKind::Inspired,
+    }
+}
+
+#[derive(Debug, Default)]
+struct HooksContext {
+    after: usize,
+    after_all: usize,
+    after_each: usize,
+    before: usize,
+    before_all: usize,
+    before_each: usize,
+}
```

Comment:
Here's a suggestion: change `usize` to `u8`

This will reduce the size of `HooksContext` by a lot. And you can add a check to not increase the counter more than `u8::MAX`. What we care more is about the counter being found.

Reply:
Another alternative is the replace the number with an enum:
```rs
enum State {
	Absent,
	Present,
	Repeated
}
```